File: src/Vograph.Core/Campus/CampusRouter.cs
using System.Text.RegularExpressions;

namespace Vograph.Core.Campus;

public static class CampusRouter
{
    public static RouteResult Find(CampusGraph graph, string fromId, string toId)
    {
        var byId = new Dictionary<string, Node>(graph.Nodes.Count, StringComparer.Ordinal);
        foreach (var node in graph.Nodes)
            byId.TryAdd(node.Id, node);

        if (fromId is null || toId is null
            || !byId.ContainsKey(fromId)
            || !byId.ContainsKey(toId))
            return RouteResult.Fail("unknown_place");

        if (fromId == toId)
            return RouteResult.Success(new Route(0, [], []));

        var adj = BuildAdjacency(graph);
        var startCost = new SearchCost(0, 0);
        var gScore = new Dictionary<string, SearchCost>(StringComparer.Ordinal) { [fromId] = startCost };
        var cameFrom = new Dictionary<string, CameFrom>(StringComparer.Ordinal);
        // Dijkstra: floor numbers alone cannot bound the cost of an authored building link.
        var open = new PriorityQueue<string, SearchPriority>();
        open.Enqueue(fromId, new SearchPriority(startCost, fromId));
        var closed = new HashSet<string>(StringComparer.Ordinal);

        while (open.TryDequeue(out var currentId, out _))
        {
            if (!closed.Add(currentId))
                continue;
            if (currentId == toId)
                return RouteResult.Success(BuildRoute(fromId, toId, byId, cameFrom, gScore[toId].Seconds));

            // A room may have several doors, but only corridors and connectors are transit places.
            if (currentId != fromId && byId[currentId].Kind == "room")
                continue;

            if (!adj.TryGetValue(currentId, out var links))
                continue;

            var currentG = gScore[currentId];
            foreach (var link in links)
            {
                if (closed.Contains(link.To))
                    continue;
                var tentative = new SearchCost(currentG.Seconds + link.Edge.Seconds,
                    currentG.Stairs + (link.Edge.Kind is "stair_up" or "stair_down" ? 1 : 0));
                if (gScore.TryGetValue(link.To, out var known) && tentative.CompareTo(known) >= 0)
                    continue;
                cameFrom[link.To] = new CameFrom(currentId, link.Edge, link.Reverse);
                gScore[link.To] = tentative;
                open.Enqueue(link.To, new SearchPriority(tentative, link.To));
            }
        }

        return RouteResult.Fail("unreachable");
    }

    public static RouteResult FindFromEntrance(CampusGraph graph, string entranceId, string roomKey)
    {
        var entrance = ResolveEntrance(graph, entranceId);
        var room = Resolve(graph, roomKey);
        if (entrance is null || room is null)
            return RouteResult.Fail("unknown_place");
        return Find(graph, entrance.Id, room.Id);
    }

    public static RouteResult FindResolved(CampusGraph graph, string fromToken, string toToken)
    {
        var from = Resolve(graph, fromToken);
        var to = Resolve(graph, toToken);
        if (from is null || to is null)
            return RouteResult.Fail("unknown_place");
        return Find(graph, from.Id, to.Id);
    }

    public static string RoomKey(string? classroomRaw)
    {
        if (string.IsNullOrWhiteSpace(classroomRaw)) return "";
        return classroomRaw.Trim().TrimEnd(';').Replace("*", "", StringComparison.Ordinal).Trim();
    }

    public static Node? Resolve(CampusGraph graph, string? token)
    {
        if (string.IsNullOrWhiteSpace(token)) return null;
        var key = RoomKey(token);
        if (key.Length == 0) return null;
        foreach (var node in graph.Nodes)
        {
            if (string.Equals(node.Id, token, StringComparison.Ordinal)
                || string.Equals(node.Id, key, StringComparison.Ordinal))
                return node;
        }
        Node? match = null;
        foreach (var node in graph.Nodes)
        {
            if (node.Room is null || !string.Equals(node.Room, key, StringComparison.Ordinal)) continue;
            if (match is not null) return null;
            match = node;
        }
        return match;
    }

    /// <summary>Schedule notation: an asterisk means УЛК; an unmarked classroom means ГК.</summary>
    public static Node? ResolveClassroom(CampusGraph graph, string? classroomRaw)
    {
        if (string.IsNullOrWhiteSpace(classroomRaw)) return null;
        var raw = classroomRaw.Trim().TrimEnd(';').Trim();
        foreach (var node in graph.Nodes)
        {
            if (string.Equals(node.Id, raw, StringComparison.Ordinal)) return node;
        }
        var building = raw.Contains("ВЦ", StringComparison.OrdinalIgnoreCase) || !raw.Contains('*') ? "ГК" : "УЛК";
        var key = RoomKey(raw);
        var vcRoom = Regex.Match(key, @"^ВЦ\s*([0-9]+[а-яa-z]?)$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
        if (vcRoom.Success) key = vcRoom.Groups[1].Value;
        Node? match = null;
        foreach (var node in graph.Nodes)
        {
            if (node.Kind != "room" || node.Building != building
                || !string.Equals(node.Room, key, StringComparison.OrdinalIgnoreCase)) continue;
            if (match is not null) return null;
            match = node;
        }
        return match;
    }

    public static Node? ResolveEntrance(CampusGraph graph, string? entranceId)
    {
        if (string.IsNullOrWhiteSpace(entranceId)) return null;
        foreach (var node in graph.Nodes)
        {
            if (string.Equals(node.Id, entranceId, StringComparison.Ordinal) && node.Kind == "entrance")
                return node;
        }
        return null;
    }

    public static Node? ChooseFrom(CampusGraph graph, string? previousRoomKey, string? lastEntranceId)
    {
        if (Resolve(graph, previousRoomKey) is { } previous)
            return previous;
        return ResolveEntrance(graph, lastEntranceId);
    }

    public static IReadOnlyList<Node> Entrances(CampusGraph graph)
    {
        List<Node> list = [];
        foreach (var node in graph.Nodes)
        {
            if (node.Kind == "entrance")
                list.Add(node);
        }
        return list;
    }

    private static Dictionary<string, List<Link>> BuildAdjacency(CampusGraph graph)
    {
        var byId = new Dictionary<string, Node>(graph.Nodes.Count, StringComparer.Ordinal);
        foreach (var node in graph.Nodes)
            byId.TryAdd(node.Id, node);

        var adj = new Dictionary<string, List<Link>>(StringComparer.Ordinal);
        foreach (var edge in graph.Edges)
        {
            if (!byId.TryGetValue(edge.From, out var from) || !byId.TryGetValue(edge.To, out var to))
                continue;
            if (CampusBlocked.WalkBlocked(edge, from, to, graph.Blocked))
                continue;
            Add(adj, edge.From, new Link(edge.To, edge, Reverse: false));
            if (!edge.OneWay)
                Add(adj, edge.To, new Link(edge.From, edge, Reverse: true));
        }
        return adj;
    }

    private static void Add(Dictionary<string, List<Link>> adj, string from, Link link)
    {
        if (!adj.TryGetValue(from, out var links))
            adj[from] = links = [];
        links.Add(link);
    }

    private static Route BuildRoute(
        string fromId,
        string toId,
        Dictionary<string, Node> byId,
        Dictionary<string, CameFrom> cameFrom,
        double seconds)
    {
        var legs = new List<Leg>();
        for (var id = toId; id != fromId;)
        {
            var step = cameFrom[id];
            legs.Add(ToLeg(step.Edge, byId[step.FromId], byId[id], step.Reverse));
            id = step.FromId;
        }
        legs.Reverse();
        return new Route((int)Math.Round(seconds, MidpointRounding.AwayFromZero), legs, []);
    }

    private static Leg ToLeg(Edge edge, Node from, Node to, bool reverse)
    {
        var kind = TravelKind(edge.Kind, reverse);
        var points = PointsFor(edge, from, to, reverse);
        if (kind == "walk")
            return new Leg(kind, from.Building, from.Floor, null, null, points);
        return new Leg(kind, from.Building, from.Floor, to.Building, to.Floor, points);
    }

    private static string TravelKind(string kind, bool reverse) => reverse
        ? kind switch
        {
            "stair_up" => "stair_down",
            "stair_down" => "stair_up",
            _ => kind
        }
        : kind;

    private static IReadOnlyList<GraphPoint> PointsFor(Edge edge, Node from, Node to, bool reverse)
    {
        if (edge.Points is not { Count: > 0 } poly)
            return [new GraphPoint(from.X, from.Y), new GraphPoint(to.X, to.Y)];
        if (!reverse)
            return poly;
        var copy = new GraphPoint[poly.Count];
        for (var i = 0; i < poly.Count; i++)
            copy[i] = poly[poly.Count - 1 - i];
        return copy;
    }

    private readonly record struct Link(string To, Edge Edge, bool Reverse);

    private readonly record struct CameFrom(string FromId, Edge Edge, bool Reverse);

    private readonly record struct SearchCost(double Seconds, int Stairs) : IComparable<SearchCost>
    {
        public int CompareTo(SearchCost other)
        {
            var seconds = Seconds.CompareTo(other.Seconds);
            return seconds != 0 ? seconds : Stairs.CompareTo(other.Stairs);
        }
    }

    private readonly record struct SearchPriority(SearchCost Cost, string NodeId) : IComparable<SearchPriority>
    {
        public int CompareTo(SearchPriority other)
        {
            var cost = Cost.CompareTo(other.Cost);
            return cost != 0 ? cost : StringComparer.Ordinal.Compare(NodeId, other.NodeId);
        }
    }
}

File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/campus/CampusRouter.kt
package ru.bgtu_voenmeh.zapara.data.campus

import java.util.PriorityQueue
import kotlin.math.roundToInt

object CampusRouter {
    fun find(graph: CampusGraph, fromId: String?, toId: String?): RouteResult {
        val byId = LinkedHashMap<String, Node>(graph.nodes.size)
        for (node in graph.nodes) byId.putIfAbsent(node.id, node)
        val start = fromId?.let { byId[it] }
        val goal = toId?.let { byId[it] }
        if (fromId == null || toId == null || start == null || goal == null) {
            return RouteResult.fail("unknown_place")
        }
        if (fromId == toId) return RouteResult.success(Route(0, emptyList()))

        val adj = buildAdjacency(graph)
        val startCost = SearchCost(0.0, 0)
        val gScore = HashMap<String, SearchCost>().apply { put(fromId, startCost) }
        val cameFrom = HashMap<String, CameFrom>()
        // Dijkstra: floor numbers alone cannot bound an authored building link's cost.
        val open = PriorityQueue<Open>(compareBy<Open> { it.cost }.thenBy { it.id })
        open.add(Open(fromId, startCost))
        val closed = HashSet<String>()

        while (true) {
            val currentId = open.poll()?.id ?: break
            if (!closed.add(currentId)) continue
            if (currentId == toId) {
                return RouteResult.success(buildRoute(fromId, toId, byId, cameFrom, gScore[toId]!!.seconds))
            }
            // A room may have several doors, but is a destination rather than a corridor.
            if (currentId != fromId && byId[currentId]!!.kind == "room") continue
            val links = adj[currentId] ?: continue
            val currentG = gScore[currentId]!!
            for (link in links) {
                if (link.to in closed) continue
                val tentative = SearchCost(currentG.seconds + link.edge.seconds,
                    currentG.stairs + if (link.edge.kind == "stair_up" || link.edge.kind == "stair_down") 1 else 0)
                val known = gScore[link.to]
                if (known != null && tentative >= known) continue
                cameFrom[link.to] = CameFrom(currentId, link.edge, link.reverse)
                gScore[link.to] = tentative
                open.add(Open(link.to, tentative))
            }
        }
        return RouteResult.fail("unreachable")
    }

    fun findFromEntrance(graph: CampusGraph, entranceId: String, roomKey: String): RouteResult {
        val entrance = resolveEntrance(graph, entranceId)
        val room = resolve(graph, roomKey)
        if (entrance == null || room == null) return RouteResult.fail("unknown_place")
        return find(graph, entrance.id, room.id)
    }

    fun findResolved(graph: CampusGraph, fromToken: String, toToken: String): RouteResult {
        val from = resolve(graph, fromToken)
        val to = resolve(graph, toToken)
        if (from == null || to == null) return RouteResult.fail("unknown_place")
        return find(graph, from.id, to.id)
    }

    fun roomKey(classroomRaw: String?): String {
        if (classroomRaw.isNullOrBlank()) return ""
        return classroomRaw.trim().trimEnd(';').replace("*", "").trim()
    }

    fun resolve(graph: CampusGraph, token: String?): Node? {
        if (token.isNullOrBlank()) return null
        val key = roomKey(token)
        if (key.isEmpty()) return null
        for (node in graph.nodes) {
            if (node.id == token || node.id == key) return node
        }
        return graph.nodes.singleOrNull { it.room != null && it.room == key }
    }

    /** Schedule notation: an asterisk means УЛК; an unmarked classroom means ГК. */
    fun resolveClassroom(graph: CampusGraph, classroomRaw: String?): Node? {
        if (classroomRaw.isNullOrBlank()) return null
        val raw = classroomRaw.trim().trimEnd(';').trim()
        graph.nodes.firstOrNull { it.id == raw }?.let { return it }
        val building = if (raw.contains("ВЦ", ignoreCase = true) || '*' !in raw) "ГК" else "УЛК"
        val cleaned = roomKey(raw)
        val key = Regex("""^ВЦ\s*([0-9]+[а-яa-z]?)$""", RegexOption.IGNORE_CASE)
            .matchEntire(cleaned)?.groupValues?.get(1) ?: cleaned
        return graph.nodes.singleOrNull {
            it.kind == "room" && it.building == building && it.room.equals(key, ignoreCase = true)
        }
    }

    fun resolveEntrance(graph: CampusGraph, entranceId: String?): Node? {
        if (entranceId.isNullOrBlank()) return null
        for (node in graph.nodes) {
            if (node.id == entranceId && node.kind == "entrance") return node
        }
        return null
    }

    fun chooseFrom(graph: CampusGraph, previousRoomKey: String?, lastEntranceId: String?): Node? {
        resolve(graph, previousRoomKey)?.let { return it }
        return resolveEntrance(graph, lastEntranceId)
    }

    fun entrances(graph: CampusGraph): List<Node> = graph.nodes.filter { it.kind == "entrance" }

    private fun buildAdjacency(graph: CampusGraph): Map<String, List<Link>> {
        val byId = LinkedHashMap<String, Node>(graph.nodes.size)
        for (node in graph.nodes) byId.putIfAbsent(node.id, node)
        val adj = HashMap<String, MutableList<Link>>()
        fun add(from: String, link: Link) {
            adj.getOrPut(from) { ArrayList() }.add(link)
        }
        for (edge in graph.edges) {
            val from = byId[edge.from] ?: continue
            val to = byId[edge.to] ?: continue
            if (walkBlocked(edge, from, to, graph.blocked)) continue
            add(edge.from, Link(edge.to, edge, reverse = false))
            if (!edge.oneWay) add(edge.to, Link(edge.from, edge, reverse = true))
        }
        return adj
    }

    private fun walkBlocked(edge: Edge, from: Node, to: Node, blocked: List<BlockedRegion>): Boolean {
        if (edge.kind != "walk" || blocked.isEmpty()) return false
        val pts = if (!edge.points.isNullOrEmpty()) edge.points
        else listOf(GraphPoint(from.x, from.y), GraphPoint(to.x, to.y))
        for (region in blocked) {
            if (region.building != from.building || region.floor != from.floor) continue
            val owner = region.ownerId
            if (owner != null && (owner == from.id || owner == to.id)) continue
            if (crosses(pts, region.left, region.top, region.right, region.bottom)) return true
        }
        return false
    }

    private fun crosses(points: List<GraphPoint>, left: Double, top: Double, right: Double, bottom: Double): Boolean {
        for (i in 0 until points.size - 1) {
            val a = points[i]
            val b = points[i + 1]
            for (s in 1 until 100) {
                val x = a.x + (b.x - a.x) * s / 100
                val y = a.y + (b.y - a.y) * s / 100
                if (x > left && x < right && y > top && y < bottom) return true
            }
        }
        return false
    }

    private fun buildRoute(
        fromId: String,
        toId: String,
        byId: Map<String, Node>,
        cameFrom: Map<String, CameFrom>,
        seconds: Double
    ): Route {
        val legs = ArrayList<Leg>()
        var id = toId
        while (id != fromId) {
            val step = cameFrom[id]!!
            legs.add(toLeg(step.edge, byId[step.fromId]!!, byId[id]!!, step.reverse))
            id = step.fromId
        }
        legs.reverse()
        return Route(seconds.roundToInt(), legs)
    }

    private fun toLeg(edge: Edge, from: Node, to: Node, reverse: Boolean): Leg {
        val kind = travelKind(edge.kind, reverse)
        val points = pointsFor(edge, from, to, reverse)
        return if (kind == "walk") Leg(kind, from.building, from.floor, null, null, points)
        else Leg(kind, from.building, from.floor, to.building, to.floor, points)
    }

    private fun travelKind(kind: String, reverse: Boolean): String = if (!reverse) kind else when (kind) {
        "stair_up" -> "stair_down"
        "stair_down" -> "stair_up"
        else -> kind
    }

    private fun pointsFor(edge: Edge, from: Node, to: Node, reverse: Boolean): List<GraphPoint> {
        val poly = edge.points
        if (poly.isNullOrEmpty()) return listOf(GraphPoint(from.x, from.y), GraphPoint(to.x, to.y))
        return if (!reverse) poly else poly.asReversed()
    }

    private data class Link(val to: String, val edge: Edge, val reverse: Boolean)
    private data class CameFrom(val fromId: String, val edge: Edge, val reverse: Boolean)
    private data class Open(val id: String, val cost: SearchCost)

    private data class SearchCost(val seconds: Double, val stairs: Int) : Comparable<SearchCost> {
        override fun compareTo(other: SearchCost): Int {
            val bySeconds = seconds.compareTo(other.seconds)
            return if (bySeconds != 0) bySeconds else stairs.compareTo(other.stairs)
        }
    }
}
